### PR TITLE
Add support missing features for flexcan on s32z27x devices

### PR DESCRIPTION
**Pull request by tunguyen4585**
This PR for updating missing features and remove the unsupported bits field

**tests/drivers/can/api**
SUITE PASS - 100.00% [can_classic]: pass = 44, fail = 0, skip = 2, total = 46 duration = 0.519 seconds
 - PASS - [can_classic.test_add_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_ext_filter] duration = 0.017 seconds
 - PASS - [can_classic.test_add_invalid_null_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_std_filter] duration = 0.015 seconds
 - PASS - [can_classic.test_bitrate_limits] duration = 0.001 seconds
 - PASS - [can_classic.test_classic_get_capabilities] duration = 0.001 seconds
 - PASS - [can_classic.test_filters_added_while_stopped] duration = 0.002 seconds
 - PASS - [can_classic.test_filters_preserved_through_bitrate_change] duration = 0.003 seconds
 - PASS - [can_classic.test_filters_preserved_through_mode_change] duration = 0.003 seconds
 - PASS - [can_classic.test_get_core_clock] duration = 0.001 seconds
 - PASS - [can_classic.test_get_state] duration = 0.001 seconds
 - PASS - [can_classic.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [can_classic.test_max_ext_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_max_std_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_receive_timeout] duration = 0.101 seconds
 - PASS - [can_classic.test_recover] duration = 0.001 seconds
 - PASS - [can_classic.test_recover_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_reject_ext_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_reject_std_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_send_and_forget] duration = 0.001 seconds
 - PASS - [can_classic.test_send_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_send_ext_id_out_of_range] duration = 0.006 seconds
 - PASS - [can_classic.test_send_fd_format] duration = 0.004 seconds
 - PASS - [can_classic.test_send_invalid_dlc] duration = 0.004 seconds
 - PASS - [can_classic.test_send_null_frame] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_ext_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_ext_id_masked] duration = 0.004 seconds
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_msgq] duration = 0.010 seconds
 - PASS - [can_classic.test_send_receive_std_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_masked] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_no_data] duration = 0.001 seconds
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_wrong_id] duration = 0.102 seconds
 - PASS - [can_classic.test_send_std_id_out_of_range] duration = 0.006 seconds
 - PASS - [can_classic.test_send_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_mode_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_state_change_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_max] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_min] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_start_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_stop_while_stopped] duration = 0.001 seconds

SUITE PASS - 100.00% [can_stats]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_stats.test_can_stats_accessors] duration = 0.001 seconds

SUITE SKIP -   0.00% [can_transceiver]: pass = 0, fail = 0, skip = 1, total = 1 duration = 0.000 seconds
 - SKIP - [can_transceiver.test_get_transceiver] duration = 0.000 seconds

SUITE PASS - 100.00% [can_utilities]: pass = 3, fail = 0, skip = 0, total = 3 duration = 0.003 seconds
 - PASS - [can_utilities.test_can_bytes_to_dlc] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_dlc_to_bytes] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_frame_matches_filter] duration = 0.001 seconds

SUITE PASS - 100.00% [canfd]: pass = 12, fail = 0, skip = 0, total = 12 duration = 0.025 seconds
 - PASS - [canfd.test_canfd_get_capabilities] duration = 0.001 seconds
 - PASS - [canfd.test_filters_preserved_through_classic_to_fd_mode_change] duration = 0.004 seconds
 - PASS - [canfd.test_filters_preserved_through_fd_to_classic_mode_change] duration = 0.004 seconds
 - PASS - [canfd.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [canfd.test_send_receive_classic] duration = 0.004 seconds
 - PASS - [canfd.test_send_receive_fd] duration = 0.003 seconds
 - PASS - [canfd.test_send_receive_mixed] duration = 0.003 seconds
 - PASS - [canfd.test_set_bitrate_data_while_started] duration = 0.001 seconds
 - PASS - [canfd.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_max] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_while_started] duration = 0.001 seconds

**tests/drivers/can/timming**
SUITE PASS - 100.00% [can_timing]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.112 seconds
 - PASS - [can_timing.test_timing] duration = 0.098 seconds
 - PASS - [can_timing.test_timing_data] duration = 0.014 seconds

**tests/drivers/can/shell**
SUITE PASS - 100.00% [can_shell]: pass = 41, fail = 0, skip = 0, total = 41 duration = 0.041 seconds
 - PASS - [can_shell.test_can_bitrate] duration = 0.001 seconds
 - PASS - [can_shell.test_can_bitrate_missing_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_bitrate_sample_point] duration = 0.001 seconds
 - PASS - [can_shell.test_can_dbitrate] duration = 0.001 seconds
 - PASS - [can_shell.test_can_dbitrate_missing_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_dbitrate_sample_point] duration = 0.001 seconds
 - PASS - [can_shell.test_can_dtiming] duration = 0.001 seconds
 - PASS - [can_shell.test_can_dtiming_missing_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_all_options] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_ext_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_ext_id_mask] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_missing_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_std_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_add_std_id_mask] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_remove] duration = 0.001 seconds
 - PASS - [can_shell.test_can_filter_remove_missing_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_combined] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_fd] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_listen_only] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_loopback] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_missing_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_normal] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_one_shot] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_raw_value] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_triple_sampling] duration = 0.001 seconds
 - PASS - [can_shell.test_can_mode_unknown] duration = 0.001 seconds
 - PASS - [can_shell.test_can_recover] duration = 0.001 seconds
 - PASS - [can_shell.test_can_recover_timeout] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_data_all_options] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_ext_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_fd] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_fd_brs] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_missing_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_no_data] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_rtr] duration = 0.001 seconds
 - PASS - [can_shell.test_can_send_std_id] duration = 0.001 seconds
 - PASS - [can_shell.test_can_show] duration = 0.001 seconds
 - PASS - [can_shell.test_can_start] duration = 0.001 seconds
 - PASS - [can_shell.test_can_stop] duration = 0.001 seconds
 - PASS - [can_shell.test_can_timing] duration = 0.001 seconds
 - PASS - [can_shell.test_can_timing_missing_value] duration = 0.001 seconds

### CODE REVIEW

File: mcux/mcux-sdk/drivers/flexcan/fsl_flexcan.c
Comment (manuargue) [inline mcux/mcux-sdk/drivers/flexcan/fsl_flexcan.c:1]: You need to submit this patch to MCUX SDK. Once approved and merged, we shall apply the same patch on this repo and note it in https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/README#L67 until the version of the SDK containing the patch is pulled.

**Review by manuargue — CHANGES_REQUESTED** [review]

**Comment by manuargue** [comment]
@tunguyen4585 there must be a zephyr pr linked to this HAL update

File: mcux/mcux-sdk/drivers/flexcan/fsl_flexcan.c
Comment (tunguyen4585) [reply mcux/mcux-sdk/drivers/flexcan/fsl_flexcan.c:1]: I updated it

**Review by mmahadevan108 — APPROVED** [review]

**Comment by mmahadevan108** [comment]
@manuargue, this change has been added to the mcux-sdk repo

File: s32/mcux/devices/S32Z270/S32Z270_device.h [hunk 1]
```diff
@@ -90,760 +90,752 @@ typedef struct {
...
-#define CAN_CS_COUNT_MB16B (63U)
+#define CAN_CS_COUNT_MB16B                         (63U)
 
 /* The count of CAN_ID */
-#define CAN_ID_COUNT_MB16B (63U)
+#define CAN_ID_COUNT_MB16B                         (63U)
 
 /* The count of CAN_WORD */
-#define CAN_WORD_COUNT_MB16B (63U)
+#define CAN_WORD_COUNT_MB16B                       (63U)
```
Comment (manuargue) [inline s32/mcux/devices/S32Z270/S32Z270_device.h:741]: pls don't introduce unnecessary reformatting on this header and keep the same format as the rest of the mcux headers for easier diff against each other, when needed

**Review by manuargue — CHANGES_REQUESTED** [review]

**Review by Dat-NguyenDuy — CHANGES_REQUESTED** [review]
Pls re-push the exact patch that was merged into mcux-sdk bitbucket. The current version in this PR is out of dated

File: s32/mcux/devices/S32Z270/S32Z270_device.h (hunk 1 above)
Comment (tunguyen4585) [reply s32/mcux/devices/S32Z270/S32Z270_device.h]: ok, I removed this commit

**Comment by tunguyen4585** [comment]
> Pls re-push the exact patch that was merged into mcux-sdk bitbucket. The current version in this PR is out of date

re-pushed it

File: s32/mcux/devices/S32Z270/S32Z270_features.h [hunk 2]
```diff
@@ -90,5 +98,17 @@
 #define FSL_FEATURE_FLEXCAN_HAS_ENHANCED_RX_FIFO (0)
 /* @brief Does not support Supervisor Mode (bitfield MCR[SUPV]. */
 #define FSL_FEATURE_FLEXCAN_HAS_NO_SUPV_SUPPORT (1)
+/* @brief Does not support Memory Error Control (bitfield MECR[HANCEI]. */
+#define FSL_FEATURE_FLEXCAN_HAS_NO_HANCEI_SUPPORT (1)
```
Comment (manuargue) [inline s32/mcux/devices/S32Z270/S32Z270_features.h:102]: ERRSR_HANCEI masks don't need to be deleted?

File: s32/mcux/devices/S32Z270/S32Z270_features.h [hunk 3]
```diff
@@ -90,5 +98,17 @@
...
+/* @brief Does not support Memory Error Control (bitfield MECR[FANCEI]. */
+#define FSL_FEATURE_FLEXCAN_HAS_NO_FANCEI_SUPPORT (1)
```
Comment (manuargue) [inline s32/mcux/devices/S32Z270/S32Z270_features.h:104]: ERRSR_FANCEI masks don't need to be deleted?

File: s32/mcux/devices/S32Z270/S32Z270_features.h [hunk 4]
```diff
@@ -90,5 +98,17 @@
...
+/* @brief Does not support Memory Error Control (bitfield MECR[CEI]. */
+#define FSL_FEATURE_FLEXCAN_HAS_NO_CEI_SUPPORT (1)
```
Comment (manuargue) [inline s32/mcux/devices/S32Z270/S32Z270_features.h:106]: ERRSR_CEI masks don't need to be deleted?

File: s32/mcux/devices/S32Z270/S32Z270_features.h (hunk 2 above)
Comment (tunguyen4585) [reply s32/mcux/devices/S32Z270/S32Z270_features.h:102]: yes, don't need to be deleted them.

File: s32/mcux/devices/S32Z270/S32Z270_features.h (hunk 3 above)
Comment (tunguyen4585) [reply s32/mcux/devices/S32Z270/S32Z270_features.h:104]: yes, don't need to be deleted them.

File: s32/mcux/devices/S32Z270/S32Z270_features.h (hunk 4 above)
Comment (tunguyen4585) [reply s32/mcux/devices/S32Z270/S32Z270_features.h:106]: yes, don't need to be deleted them.

**Review by Dat-NguyenDuy — APPROVED** [review]

**Review by manuargue — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**